src/Microsoft.OData.Core/UriParser/Parsers/FunctionOverloadResolver.cs

```diff
@@ -217,45 +216,58 @@ internal static bool ResolveOperationFromList(string identifier, IEnumerable<str
                     throw ExceptionUtil.CreateBadRequestError(ODataErrorStrings.RequestUriProcessor_SegmentDoesNotSupportKeyPredicates(identifier));
                 }
 
-                matchingOperation = candidateMatchingOperations.Single();
+                matchingOperation = candidatesMatchingOperations.Count > 0 ? candidatesMatchingOperations[0] : null;
                 return true;
             }
 
-            if (foundActionsWhenLookingForFunctions.Count > 0)
+            if (foundActionsWhenLookingForFunctions != null && foundActionsWhenLookingForFunctions.Count > 0)
             {
                 throw ExceptionUtil.CreateBadRequestError(ODataErrorStrings.RequestUriProcessor_SegmentDoesNotSupportKeyPredicates(identifier));
             }
 
             // If more than one overload matches, try to select based on optional parameters
-            if (candidateMatchingOperations.Count() > 1)
+            if (candidatesMatchingOperations.Count > 1)
             {
-                candidateMatchingOperations = candidateMatchingOperations.FindBestOverloadBasedOnParameters(parameterNames);
+                candidatesMatchingOperations = candidatesMatchingOperations.FindBestOverloadBasedOnParameterCount(parameterNames.Count);
             }
 
-            if (candidateMatchingOperations.Count() > 1)
+            if (candidatesMatchingOperations.Count > 1)
             {
                 throw new ODataException(ODataErrorStrings.FunctionOverloadResolver_NoSingleMatchFound(identifier, string.Join(",", parameterNames.ToArray())));
             }
 
-            matchingOperation = candidateMatchingOperations.SingleOrDefault();
+            matchingOperation = candidatesMatchingOperations.Count > 0 ? candidatesMatchingOperations[0] : null;
             return matchingOperation != null;
         }
 
         internal static bool HasAny<T>(this IEnumerable<T> enumerable) where T : class
         {
-            IList<T> list = enumerable as IList<T>;
-            if (list != null)
+            if (enumerable != null && enumerable.GetEnumerator().MoveNext())
             {
-                return list.Count > 0;
+                return true;
             }
 
-            T[] array = enumerable as T[];
-            if (array != null)
+            return false;
+        }
+
+        internal static bool CountGreaterThan<T>(this IEnumerable<T> enumerable, int count) where T : class

```

Comment:
>CountGreaterThan [](start = 29, length = 16)

This assume you can enumerate the enumerable multiple times (without copying).  works for a list, but may not work for other enumerables.

Reply:
This is not intended to be used for lists. This is intended for Enumerable, here is the implementation of count in Linq, this is an improvment only for IEnumerables ->
        public static int Count<TSource>(this IEnumerable<TSource> source)
        {
            if (source == null)
            {
                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
            }

            if (source is ICollection<TSource> collectionoft)
            {
                return collectionoft.Count;
            }

            if (source is IIListProvider<TSource> listProv)
            {
                return listProv.GetCount(onlyIfCheap: false);
            }

            if (source is ICollection collection)
            {
                return collection.Count;
            }

            int count = 0;
            using (IEnumerator<TSource> e = source.GetEnumerator())
            {
                checked
                {
                    while (e.MoveNext())
                    {
                        count++;
                    }
                }
            }

            return count;
        }
 

---
In reply to: [370855165](https://github.com/OData/odata.net/pull/1602#discussion_r370855165) [](ancestors = 370855165)